### PR TITLE
Implement updated API design

### PR DESCRIPTION
**Pull request by aashikam**
Fixes https://github.com/wso2-ballerina/module-googlespreadsheet/issues/70

https://developers.google.com/apps-script/reference/spreadsheet

### CODE REVIEW

File: src/gsheets4/data_mappings.bal [hunk 1]
```diff
@@ -14,35 +14,35 @@
...
-function convertToSpreadsheet(json jsonSpreadsheet) returns Spreadsheet {
-    Spreadsheet spreadsheet = new;
-    spreadsheet.spreadsheetId = jsonSpreadsheet.spreadsheetId.toString();
-    json|error spreadsheetProperties= jsonSpreadsheet.properties;
-    spreadsheet.properties = !(spreadsheetProperties is error)
-                             ? convertToSpreadsheetProperties(spreadsheetProperties) : {};
-    spreadsheet.spreadsheetUrl = jsonSpreadsheet.spreadsheetUrl.toString();
-    spreadsheet.sheets = !(jsonSpreadsheet.sheets is error)
-                         ? convertToSheets(<json[]> jsonSpreadsheet.sheets) : [];
+import ballerina/ 'lang\.int as ints;
```
Comment (daneshk) [inline src/gsheets4/data_mappings.bal]:
```suggestion
import ballerina/lang.'int as ints;
```

File: src/gsheets4/data_mappings.bal [hunk 2]
```diff
@@ -14,35 +14,35 @@
...
 
+function convertToSpreadsheet(json jsonSpreadsheet, Client spreadsheetClient) returns Spreadsheet {
+    string id = jsonSpreadsheet.spreadsheetId.toString();
+    json | error spreadsheetProperties = jsonSpreadsheet.properties;
+    SpreadsheetProperties properties = !(spreadsheetProperties is error)
+    ? convertToSpreadsheetProperties(spreadsheetProperties) : {};
+    string spreadsheetUrl = jsonSpreadsheet.spreadsheetUrl.toString();
+    Sheet[] sheets = !(jsonSpreadsheet.sheets is error)
```
Comment (daneshk) [inline src/gsheets4/data_mappings.bal]: If there is an error, shall we log the error and return an empty value

File: src/gsheets4/sheet.bal [hunk 3]
```diff
@@ -0,0 +1,493 @@
...
+                }
+            ]
+        };
+        string renamePath = SPREADSHEET_PATH + PATH_SEPARATOR + self.parentId + BATCH_UPDATE_REQUEST;
+        json | error response = sendRequestWithPayload(self.httpClient, renamePath, payload);
+        if (response is error) {
+            return response;
+        } else {
+            self.properties.title = name;
+            return ();
```
Comment (daneshk) [inline src/gsheets4/sheet.bal]:
```suggestion
```

File: src/gsheets4/sheet.bal [hunk 4]
```diff
@@ -0,0 +1,493 @@
...
+                }
+            ]
+        };
+        string deleteSheetPath = SPREADSHEET_PATH + PATH_SEPARATOR + self.parentId + BATCH_UPDATE_REQUEST;
+        json | error response = sendRequestWithPayload(self.httpClient, deleteSheetPath, payload);
+        if (response is json) {
+            return;
+        } else {
+            return response;
+        }
```
Comment (daneshk) [inline src/gsheets4/sheet.bal]: Shall we change the return logic to, as we don't need to return when success
```ballerina
if (response is error) {
    return response;
}
```

Comment (daneshk) [reply src/gsheets4/sheet.bal]: please change all places

File: src/gsheets4/spreadsheet.bal [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+    }
+
+    # Get sheets of the spreadsheet.
+    # + sheetName - Name of the sheet to retrieve
+    # + return - Sheet object on success and error on failure
+    public function getSheetByName(string sheetName) returns Sheet | error {
+        Sheet[] sheets = self.sheets;
+        error err = error(SPREADSHEET_ERROR_CODE, message = "Sheet not found");
+        if (sheets.length() == 0) {
+            return err;
```
Comment (daneshk) [inline src/gsheets4/spreadsheet.bal]: IMO, we can get rid of this check. if `sheets` is an empty array, we don't get inside foreach

File: src/gsheets4/spreadsheet.bal [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+                }
+            ]
+        };
+        string deleteSheetPath = SPREADSHEET_PATH + PATH_SEPARATOR + self.spreadsheetId + BATCH_UPDATE_REQUEST;
+        json | error response = sendRequestWithPayload(self.httpClient, deleteSheetPath, payload);
+        if (response is error) {
+            return response;
+        } else {
+            self.properties.title = name;
+            return;
```
Comment (daneshk) [inline src/gsheets4/spreadsheet.bal]:
```suggestion
```

File: src/gsheets4/utils.bal [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+
+function setResponseError(json jsonResponse) returns error {
+    error err = error(SPREADSHEET_ERROR_CODE, message = jsonResponse.toString());
+    return err;
+}
+
+function setResError(error errorResponse) returns error {
+    error err = error(SPREADSHEET_ERROR_CODE, message = <string>errorResponse.detail()?.message);
+    return err;
+}
```
Comment (daneshk) [inline src/gsheets4/utils.bal]: Shall we combine these two functions like below,
```ballerina
function setResponseError(json|error errorResponse) returns error {
  if (errorResponse is json) {
        return error(SPREADSHEET_ERROR_CODE, message = errorResponse.toString());
  } else {
        return error(SPREADSHEET_ERROR_CODE, message = <string>errorResponse.detail()?.message);
  }
}
```

File: src/gsheets4/utils.bal [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+        }
+    } else {
+        return createConnectorError(httpResponse);
+    }
+}
+
+function createConnectorError(error errorResponse) returns error {
+    error err = error(SPREADSHEET_ERROR_CODE, message = <string>errorResponse.detail()?.message);
+    return err;
+}
```
Comment (daneshk) [inline src/gsheets4/utils.bal]: IMO, we don't need this function as we can use the `getResponseError` function

File: src/gsheets4/utils.bal [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+    if (value is int) {
+        return value;
+    } else if (value is float) {
+        return value;
+    } else {
+        return value.toString();
+    }
+}
+
+function setJsonResponse(json jsonResponse, int statusCode, Client cli) returns Spreadsheet | error {
```
Comment (daneshk) [inline src/gsheets4/utils.bal]:
```suggestion
function fromJsonResponse(json jsonResponse, int statusCode, Client cli) returns Spreadsheet | error {
```

File: src/gsheets4/utils.bal [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+        return convertToSpreadsheet(jsonResponse, cli);
+    }
+    return setResponseError(jsonResponse);
+}
+
+function validateStatusCode(json response, int statusCode) returns error? {
+    if (statusCode != http:STATUS_OK) {
+        return setResponseError(response);
+    }
+    return ();
```
Comment (daneshk) [inline src/gsheets4/utils.bal]:
```suggestion
```

File: src/gsheets4/utils.bal [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+    }
+    return setResponseError(jsonResponse);
+}
+
+function validateResponse(json jsonResponse, int statusCode, Client cli) returns Spreadsheet | error {
+    if (statusCode == http:STATUS_OK) {
+        return convertToSpreadsheet(jsonResponse, cli);
+    }
+    return setResponseError(jsonResponse);
+}
```
Comment (daneshk) [inline src/gsheets4/utils.bal]: IMO, we can remove this function and use `fromJsonResponse` instead

File: src/gsheets4/utils.bal [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+        return setResponseError(response);
+    }
+    return ();
+}
+function setResponse(json jsonResponse, int statusCode) returns error? {
+    if (statusCode == http:STATUS_OK) {
+        return ();
+    }
+    return setResponseError(jsonResponse);
+}
```
Comment (daneshk) [inline src/gsheets4/utils.bal]: IMO, we can remove this function and use `validateStatusCode` instead

Comment (aashikam) [reply src/gsheets4/utils.bal]: Fixed

File: src/gsheets4/utils.bal (hunk 11 above)
Comment (aashikam) [reply src/gsheets4/utils.bal]: Fixed

File: src/gsheets4/utils.bal (hunk 10 above)
Comment (aashikam) [reply src/gsheets4/utils.bal]: Fixed

File: src/gsheets4/utils.bal (hunk 9 above)
Comment (aashikam) [reply src/gsheets4/utils.bal]: Fixed

File: src/gsheets4/utils.bal (hunk 8 above)
Comment (aashikam) [reply src/gsheets4/utils.bal]: Fixed

File: src/gsheets4/utils.bal (hunk 7 above)
Comment (aashikam) [reply src/gsheets4/utils.bal]: Fixed

File: src/gsheets4/spreadsheet.bal (hunk 6 above)
Comment (aashikam) [reply src/gsheets4/spreadsheet.bal]: Fixed

File: src/gsheets4/spreadsheet.bal (hunk 5 above)
Comment (aashikam) [reply src/gsheets4/spreadsheet.bal]: Fixed

File: src/gsheets4/sheet.bal (hunk 4 above)
Comment (aashikam) [reply src/gsheets4/sheet.bal]: Fixed

File: src/gsheets4/sheet.bal (hunk 3 above)
Comment (aashikam) [reply src/gsheets4/sheet.bal]: Fixed

File: src/gsheets4/data_mappings.bal (hunk 2 above)
Comment (aashikam) [reply src/gsheets4/data_mappings.bal]: Fixed

File: src/gsheets4/data_mappings.bal (hunk 1 above)
Comment (aashikam) [reply src/gsheets4/data_mappings.bal]: Fixed

**Pull request merged**